rfcs/015 - Reworked Compiler Front End.md

```diff
@@ -0,0 +1,179 @@
+
+The new compiler front-end will also be able to determine function signatures without requiring type annotations on parameters or the return type.
+A simple example might look like:
+```ln
+fn add4(num) {
+  return add(num, 4);
+}
+```
+
+At the time of writing this RFC, this function fails to parse due to a missing type annotation on `num`, but it's intended to support gradual typing even for function parameters in Alan.
+The current front-end will require substantial work to support this, requiring one of:
+- An interface (internal or in `root.ln`) that ensures there's an `add` operation for the type of `num` (for example, `Addable`).
+- Type-checking can only happen once the function is inlined.
+
+However, the new type resolution system would simply add a constraint that there be a function named `add` that satisfies the signature `(T, 4 as T) -> T` (ignoring the fact that arithmetic operations in Alan actually use `Result` values).
+This wouldn't require the `Addable` interface to be defined at all, and instead would allow for an anonymous interface that's only visible in the compiler to be created.
+The constraints would then not need to be computed for each time the function is called - instead, we must only ensure that the value passed in to `add4` satisfies the so-called "Addable" interface.
+
+A more complex example would be the `toString` implementation for `Maybe`, which could be rewritten as such under the new front-end:
+```ln
+fn toString(val) {
+  if val.isSome() {
+    return val.getMaybe().toString();
+  } else {
+    return 'none';
+  }
+}
+```
+
+In order to share more nuanced details about the new front-end, here are the steps the compiler might take upon having to use this function:
+1. Assign the variable `val` to type `T`. There are no constraints on the type `T`.
+2. Desugar the if/else. How this is done is unimportant, but the body of the `toString` function will effectively be desugared into (ignore the invalid `conditionTailFn` function body - it's required for `evalcond` to work but its contents won't be checked by the type checker, since this desugaring will actually happen just before generating the Microstatements):
+```ln
+const conditionalTable = newarr(2);
+const condition1Boolean = val.isSome();
+const condition1ThenFn = fn(): U { return val.getMaybe().toString(); };
+condfn(conditionalTable, condition1Boolean, condition1ThenFn);
+const condition2Boolean = true;
+const condition2ThenFn = fn(): U { return 'none'; };
+condfn(conditionalTable, condition2Boolean, condition2ThenFn);
+const conditionTailFn = fn(): U {};
+const conditionResult: Maybe<U> = evalcond(conditionalTable, conditionTailFn);
+const conditionResultUnwrapped: U = conditionResult.getMaybe();
+return conditionResultUnwrapped;
+```
+3. Add the constraint that `val` has type `A` where there's a function called `isSome` with the signature `(A) -> bool`.
```

Comment:
I also don't see how the `(A) -> bool` is determined here without some deep knowledge of how conditionals work. That could be fine, but wouldn't it also function it just determined a signature `(A) -> X` where X is just some other type, and it just happens to be `bool` since there is no other implementation of `isSome` in the standard library.

Reply:
I guess it would be `(A) -> X` where `X` is constrained to be a `bool`, I left that part out since it's part of the same step but I'll clarify.